### ISSUE TITLE
visit_concept_id

### ISSUE THREAD

**Issue by chrisroederucdenver** [cladteam/CCDA_OMOP_by_Python#72]
Determining the OMOP visit_concept_id that describes the type of visit in basic terms as in-patien or out-patient etc is challenging.
[detail on google](https://docs.google.com/document/d/1gp1mI83Ee2A3z3a6YxaVMUTKcXIKe6buCxGZMelN3_g/edit)

- [x] waiting on #149 for where the VSAC json file is on foundry.
- [ ] the mapping

**Comment by chrisroederucdenver** [comment]
Steph suggested we look for ACT-enc encounter codes and not consider a mapping from those CPT code to OMOP visit...

From Steph: https://terminology.hl7.org/ValueSet-encounter-class.html

**Comment by chrisroederucdenver** [comment]
IN THE DATA....

- encompassingEncounter doesn't have codes that make sense. encounters do.
- Most files have both and they overlap. ToC only has an encompassingEncounter and an SE.
- Some files have a very large SE window (decades) along with one or more encounters, those don't have encompassingEncounter elements.
- ToC is an oddball in that it doesn't have encounters, has an encompassingEncounter and a very large SE range.
- bennis is the opposite oddball that has many encounters, no encompassingEncounter and a very large SE range.

SUGGESTION:
- create a visit for each encounter.  assume it is The One.
  - check somehow for concordance with encompassingEncounter? The reason ot use the encounter is it has a source for visit_concept_id
  - If there are many encounters, create one for each, and POST PROCESS by dates to assign the FK??? (after using the last ...b/c that's what will populate the PK)
    - based on the sample the wide date range in the serviceEvent is a clue to having many encounters.
  - watchout for DATES THAT DON"T PARSE: empty strings and "UNK"

**Comment by chrisroederucdenver** [comment]
Two sets of concepts in github.com/cladteam/CCDA_OMOP_Private:

[EncounterTypeCode.csv](https://github.com/chrisroederucdenver/CCDA_OMOP_Private/blob/main/EncounterTypeCode.csv
) The value set of codes that describe encounters in CCDA.
[CPT_visit_codes.csv](https://github.com/chrisroederucdenver/CCDA_OMOP_Private/blob/main/CPT_visit.csv) a list of concepts from Athena that are in the CPT4 vocabulary and have domain_id of 'visit'

They seem to be mutually exclusive.

**Comment by chrisroederucdenver** [comment]
I'm downloading CPT4 on my laptop and will be able to check concept_relationship for mappings for the VSAC value set when it downloads. For now in our test data we've seen 99212 and 99213 that both map to the standard concept in the Visit vocabulary (that has visit doamin_id) 9202 for Outpatient Visit.

**Comment by chrisroederucdenver** [comment]
in "Analysis Needed" because the valueset for the code field in an EncounterActivity is huge and largely populated with procedures, some of which give a strong clue about the visit type, but many don't. This also relates to mappings in the OMOP concept_relationship table that maps some of them to visit, and rumors that those mappings will be removed.

For the moment, in our test data, as commented above, we have concepts that map to OP. Missing from the comment are other concepts that would map to IP but don't.

**Comment by stephanieshong** [comment]
@chrisroederucdenver  @croeder   FYI -  the classCode attribute (right next to the moodCode attribute) was more for the CCDA xlm document that I was looking through not FHIR resources. 
The ActEnc valueset which all about the encounter and visit_cocnept_id. Here in CCDA, for [Encounter Activity](https://www.hl7.org/ccdasearch/templates/2.16.840.1.113883.10.20.22.4.49.html) there is a class code and it is restricted to [ActClass ](https://terminology.hl7.org/1.0.0/CodeSystem-v3-ActClass.html). It may have a value of 'ENC' or many others. In fact, if you read the Encounter Activity page, I think it says the ActClass MUST be ENC.

**Comment by chrisroederucdenver** [comment]
@mbangudi  @richard1933  @stephanieshong @jmcmurry 
We should consider SSSOM to document the provenance of vocabulary mappings.
https://academic.oup.com/database/article/doi/10.1093/database/baac035/6591806
https://mapping-commons.github.io/sssom/

**Comment by chrisroederucdenver** [comment]
@mbangudi @richard1933 @tannerzhang 
FYI VSAC data is parsed and located here: /All of Us-cdb223/Identified: HIN - HIE/CCDA/transform/vsac_EncounterTypeCode

**Comment by stephanieshong** [comment]
@chrisroederucdenver  - In CCDA data provide you with the place of visit within the encounter tag? If so could we use that value to capture the visit_concept?

**Comment by chrisroederucdenver** [comment]
@stephanieshong @tannerzhang 
Another valueset from VSAC, in CSV This one for HSLOC
[HealthcareServiceLocation.csv](https://github.com/user-attachments/files/18157882/HealthcareServiceLocation.csv)

**Comment by stephanieshong** [comment]
@chrisroederucdenver - will you be able to consistently capture the place of service like below?
< ...
code code="1160-1" codeSystem="2.16.840.1.113883.6.259" codeSystemName="HL7 HealthcareServiceLocation" displayName="Urgent Care Center"/>

**Comment by stephanieshong** [comment]
If the data is able to give you consistenly the place of health service location via the
 codeSystem="2.16.840.1.113883.6.259" codeSystemName="HL7 HealthcareServiceLocation"

then certainly we can map the  visit_concept_id  using this value: 

Code          Ch 15: NHSN Healthcare Service Location Code
Would you be able to confirm this is the case in your sample CCDA dataset?

**Comment by chrisroederucdenver** [comment]
Probably not for every document. When I found “SDLOC”,  they had these codes. There are some very weird files in our sample set.
Like a lot of things, we’ll have a better idea when the real data starts flowing. We get encounters that don’t have location data in the participant element…that don’t have an participant element.

From: Stephanie S. Hong ***@***.***>
Date: Monday, December 16, 2024 at 3:56 PM
To: cladteam/CCDA_OMOP_by_Python ***@***.***>
Cc: Roeder, Chris ***@***.***>, Mention ***@***.***>
Subject: Re: [cladteam/CCDA_OMOP_by_Python] visit_concept_id (Issue #72)

@chrisroederucdenver<https://github.com/chrisroederucdenver> - will you be able to consistently capture the place of service like below?
< ...
code code="1160-1" codeSystem="2.16.840.1.113883.6.259" codeSystemName="HL7 HealthcareServiceLocation" displayName="Urgent Care Center"/>

**Comment by stephanieshong** [comment]
Would you be able to provide a list of code/codeSystem (in order)  you will be able to use to map the encounter types?

For example for encounter type we can go down the list of coded elements that you can chase after and 
COALESCE() the non- null coded values in the clinicalDocumet header that you want to use to encode the encounter type. 
And then give us the list of coded value/code System to map  to visite_concept_id ( codeSystem will be oid for a known vocabulary Id that we can identify as the well known externally managed terminologySystem or valueSet that we can reference and parse to identify the correct concept id) 

related ACTS connected to the CDA document you will have ClincialDocument class that tells you the specific service performed and the encounter in which they took place.  This is the enformation we need to assign the encouter type, i.e. visit_concept_id 

Important portion you want extracted is the following section: 
CDAHeader-> provider, the type of service and the location of service ( often in LOINC classification with type of document, type of service, and the author and location of service provided. 
**<tag of the performer or the author role - common service provided by a health care professional may include, care history, physicals, consults, operation or procedures, dishcharge or transfer, progress notes, lab reports, imaging studies, etc. 
**<tag professional level of the provider** - provider level of provider can be  physicians, nurses, or residents, etc. 
**<tag encompassingEncounterlocation act location of the participation class** - location of serivce which may include, hospital, home, hospice, long term care facility, Emergency Department or Ambulartory care. 
 

There are complexity that get added with relatedDocuments, but we will worry about that later. 
Hope this helps.

**Comment by chrisroederucdenver** [comment]
We need to flesh out the locations in the document before we can identify valuesets. I have a rough idea about what  you describe, but need to spend time to flesh it out and understand it. When I do so, the spec. will list value sets and I will pass them on.

**Comment by stephanieshong** [comment]
Thank you. We do not wish to hap hazardly map data elements unnecessarily, as this process requires significant time and effort for thorough review. Hope you understand the level of time committment required for mapping these data elements, espcially if the value is not coded with proper terminology.

**Comment by tursynay** [comment]
Added Steph to assignee as she has been working on this ticket too

**Comment by chrisroederucdenver** [comment]
@stephanieshong @tursynay

**Comment by chrisroederucdenver** [comment]
@stephanieshong @tursynay 
Moving to Jan. 8 because that's far more realistic than Dec. 11 (last week), and especially for this ticket, the expectations of it involving a process of cooperation to figure this out are much higher and more flexible.

Jan. 8 is not a do-or-die date either! We may have enough to populate this field in short order* as Stephanie has some great suggestions regarding looking at data available from site and providers that are related to the encounter in many ways. Our methods will improve, and improve even more when we get real data. I think we should have at least a basic version together by the end of Jan, with progress to report along the way. 

* re:short order.
We have identified two locations and associated valuesets that we can apply this too. The vocabulary mappings are likely messy, and if I understand Steph, she's suggesting a mapping that goes from a coalesce of values from multiple sources, and involves some real effort.  Where a normal "coalesce" function chooses the first non-null value from a list, I could see this mapping being more like choosing the first workable translation to visit_concpet_id: workable rather than null.

**Comment by stephanieshong** [comment]
For each patient encounter you will need to logically decide which CCDA data element you will use to decide on the visit concept ids and give us the distinct list of encounter type you want us to map. We cannot make the logical decision for you as you will be parsing through the CCDA data elements. If you can construct the list of information with the following : pt/ encounter / encounter date/ encounter class code that let you know that this encounter occurred/ place of service/  then we should be able to assign a visit_concept_id base on this. 
@chrisroederucdenver - let me know if this helps.

**Comment by stephanieshong** [comment]
As we discussed during our meeting, we can begin by reviewing the list of coded values I mentioned earlier as a starting point. This will help us determine the appropriate order for the COALESCE() function. If you can build the dataset with the key columns we need to review, we can work together to decide on the order for assigning the visit_concept_id. Without examining the data first, it will be challenging to make an informed decision. 
Once we finalize the order to assign the visit_concept_id, you can then implement the logic in your COALESCE() function.

We need to assign a visit_concept_id for a given pt/encounter but currently we do not know what data you are getting in your sample data that is not null.
So lets gather that first and go from there.
Does this make sense?

**Comment by stephanieshong** [comment]
Lets review the dataset you are extracting from CCDA together so we can have a concrete plan as to how to decide on the visit_concept_id

**Comment by tursynay** [comment]
Tanner and Steph are trying to crate a shortlist of visit types from OMOP vocab that we can target as visit concept ID. This will be based on the codeset that Chris will pull out of CCDA files. This shortlist will be something like OP, IP, ERIP, Non-hospital institution visit (hospice, psychiatric ward, skilled nursing, etc.)

**Comment by stephanieshong** [comment]
@tannerzhang - yesterday we identified the list of codes that can be use to map visit_concept_id. Do you want to go over the list of codes which we can use to create the visit_concept_xwalk mapping table?

**Comment by tannerzhang** [comment]
Do we have a finalized short list of codes? @stephanieshong

**Comment by chrisroederucdenver** [comment]
@tannerzhang  see my comment on #187 that links to a workbook that filters the concepts used by the paths.

**Comment by tannerzhang** [comment]
[NHSN-Healthcare-Facility-Patient-Care-Locations-Chapter15.csv](https://github.com/user-attachments/files/18380214/NHSN-Healthcare-Facility-Patient-Care-Locations-Chapter15.csv)
@stephanieshong asked me to create the visit_concept_xwalk_mapping table( fusion sheet) to create the visit_concept_xwalk dataset in the same mapping table reference folder. Working on this now.

**Comment by tannerzhang** [comment]
Draft table created for review https://foundry.cladplatform.org/workspace/fusion/spreadsheet/ri.fusion.main.document.9bb3ea91-7c82-4d9b-b2e9-44d0026d7ff1/spreadsheet/08842222866857352

**Comment by tannerzhang** [comment]
Known issue: currently miss mappings for MISCELLANEOUS AREAS , COMMUNITY LOCATIONS, NON-PATIENT CARE LOCATIONS, HEALTHCARE PERSONNEL SAFETY CDA. 
But we are in good shape for inpatient/outpatient visits

**Comment by tursynay** [comment]
@chrisroederucdenver to do the next step and send it back to @tannerzhang to review.

**Comment by chrisroederucdenver** [comment]
comments in #187  
Looks like we'll have to consider the HSLOC code in participant and then in some cases the CPT-4 code in the main encounter part. Look out, for now, HSLOC is in a visit mapping table and the CPT-4 to visit mappings are in the ccda_value_set_mapping_table_dataset.

**Comment by tannerzhang** [comment]
Worked on this together and reviewed with @chrisroederucdenver . With the understanding to rerun the snooper and do necessary new mappings, the ticket should be good to be closed.

99213
99212
99412
99222
99241
99221 Added from CPT for encounter.

**Comment by chrisroederucdenver** [comment]
upgraded error messages and the bin/compare_output.sh script, completed the conversion of mapping to use the foundry datasets, added a tool for looking up concepts in the 3 vocab maps.

We still have to get encompassingEncouters done, but that's another ticket. #198
